pytorch_lightning/utilities/hparams_mixin.py

```diff
@@ -0,0 +1,180 @@
+        if not frame:
+            frame = inspect.currentframe().f_back
+        save_hyperparameters(self, *args, ignore=ignore, frame=frame)
+
+    def _set_hparams(self, hp: Union[dict, Namespace, str]) -> None:
+        hp = self._to_hparams_dict(hp)
+
+        if isinstance(hp, dict) and isinstance(self.hparams, dict):
+            self.hparams.update(hp)
+        else:
+            self._hparams = hp
+
+    @staticmethod
+    def _to_hparams_dict(hp: Union[dict, Namespace, str]):
+        if isinstance(hp, Namespace):
+            hp = vars(hp)
+        if isinstance(hp, dict):
+            hp = AttributeDict(hp)
+        elif isinstance(hp, PRIMITIVE_TYPES):
+            raise ValueError(f"Primitives {PRIMITIVE_TYPES} are not allowed.")
+        elif not isinstance(hp, ALLOWED_CONFIG_TYPES):
+            raise ValueError(f"Unsupported config type of {type(hp)}.")
+        return hp
+
+    @property
+    def hparams(self) -> Union[AttributeDict, dict, Namespace]:
+        if not hasattr(self, "_hparams"):
+            self._hparams = AttributeDict()
+        return self._hparams
+
+    @property
+    def hparams_initial(self) -> AttributeDict:
+        if not hasattr(self, "_hparams_initial"):
+            return AttributeDict()
+        # prevent any change
+        return copy.deepcopy(self._hparams_initial)
+
+    @hparams.setter
+    def hparams(self, hp: Union[dict, Namespace, Any]):
+        hparams_assignment_name = self.__get_hparams_assignment_variable()
+        self._hparams_name = hparams_assignment_name
+        self._set_hparams(hp)
+        # this resolves case when user does not uses `save_hyperparameters` and do hard assignement in init
+        if not hasattr(self, "_hparams_initial"):
+            self._hparams_initial = copy.deepcopy(self._hparams)
+
+    def __get_hparams_assignment_variable(self):
+        """
+        looks at the code of the class to figure out what the user named self.hparams
+        this only happens when the user explicitly sets self.hparams
+        """
+        try:
+            class_code = inspect.getsource(self.__class__)
+            lines = class_code.split("\n")
+            for line in lines:
+                line = re.sub(r"\s+", "", line, flags=re.UNICODE)
+                if ".hparams=" in line:
+                    return line.split("=")[1]
+        except Exception:
+            return "hparams"
+
+        return None
+
+
+def merge_hparams(lightning_hparams: dict, data_hparams: dict) -> dict:
+    """Merge the hparams of a LightningModule and a LightningDataModule and return them.
+
+    If there is an overlap between the hparams in the LightningModule and in the
+    LightningDataModule an exception is raised.
+
+    Args:
+        lightning_hparams: the hyperparameters of a LightningModule
+        data_hparams: the hyperparameters of a LightningDataModule
+    """
+    colliding_keys = [key for key in data_hparams.keys() if key in lightning_hparams]
```

Comment:
You can use the intersection of `dict_keys`

```suggestion
    colliding_keys = lightning_hparams.keys() & data_hparams.keys()
```